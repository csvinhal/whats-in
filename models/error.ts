export interface ErrorResponse {
    title: string
    subtitle: string
}
