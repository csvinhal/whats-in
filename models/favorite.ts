import mongoose, { model, models } from 'mongoose'

interface IFavorite {
    slug: string
}

const FavoriteSchema = new mongoose.Schema({
    slug: {
        type: String,
        index: true,
        unique: true,
    },
})

export default models.Favorite ||
    model<IFavorite & mongoose.Document>('Favorite', FavoriteSchema)
