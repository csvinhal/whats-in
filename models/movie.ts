export interface Movie {
    title: string
    imdb: string
    year: string
    poster: string
    favorite: boolean
}

export interface MovieResponse {
    movies?: Movie[]
    totalResults?: string
}

export interface SearchResponse {
    Title: string
    Year: string
    imdbID: string
    Poster: string
}

export interface OmdbMovieResponse {
    Search?: SearchResponse[]
    totalResults?: string
    Error?: string
    Response: 'True' | 'False'
}
