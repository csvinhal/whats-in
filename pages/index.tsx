import cx from 'classnames'
import EmptyState from '../components/EmptyState/EmptyState'
import ErrorState from '../components/ErrorState/ErrorState'
import Grid from '../components/Grid/Grid'
import GridItem from '../components/GridItem/GridItem'
import Layout from '../components/Layout/Layout'
import MovieList from '../components/MovieList/MovieList'
import MoviesLoading from '../components/MoviesLoading/MoviesLoading'
import SearchBar from '../components/SearchBar/SearchBar'
import { useFetchMovies } from '../hooks/useFetchMovies'

export const Home = (): JSX.Element => {
    const {
        state: { loading, data, error },
        effectFetchMovies,
    } = useFetchMovies()

    let content
    if (loading) content = <MoviesLoading />
    else if (error)
        content = <ErrorState title={error.title} subtitle={error.subtitle} />
    else if (data) content = <MovieList movies={data.movies} />
    else
        content = (
            <Grid>
                <GridItem xs>
                    <EmptyState
                        title="Don't know what to search?"
                        subtitle="Here's an offer you can't refuse"
                    />
                </GridItem>
            </Grid>
        )

    return (
        <Layout className="container mb-3">
            <Grid>
                <GridItem xs>
                    <div
                        className={cx('container__search mb-4 ', {
                            'container__search--is-empty-state':
                                !loading && !data,
                        })}
                    >
                        <SearchBar
                            placeholder="Search movies..."
                            onChange={e => {
                                effectFetchMovies(e?.target?.value)
                            }}
                        />
                    </div>
                </GridItem>
            </Grid>
            {content}

            <style jsx>{`
                .container__search--is-empty-state {
                    margin-bottom: 25vh;
                }
            `}</style>
        </Layout>
    )
}

export default Home
