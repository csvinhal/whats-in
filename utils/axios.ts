import axios from 'axios'

const axiosInstance = axios.create({
    baseURL: process.env.baseUrl,
})

export default axiosInstance
