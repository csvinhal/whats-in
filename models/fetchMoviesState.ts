import { MovieResponse } from './movie'

export interface FetchMoviesError {
    title: string
    subtitle: string
}

export interface FetchMoviesState {
    loading: boolean
    data: MovieResponse
    error: FetchMoviesError
}
