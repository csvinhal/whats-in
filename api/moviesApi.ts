import { MovieResponse } from '../models/movie'
import { MovieDetails } from '../models/movieDetails'
import axios from '../utils/axios'

export const getMovies = (title: string) => {
    return axios.get<MovieResponse>(`/movies`, {
        params: { title },
    })
}

export const getMovieDetail = (slug: string) => {
    return axios.get<MovieDetails>(`/movie-detail/${slug}`)
}

export const updateMovieFavorite = (slug: string, favorite: boolean) => {
    return axios.post<MovieDetails>(`/movie-detail/${slug}`, { favorite })
}
