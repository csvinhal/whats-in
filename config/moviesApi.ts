import { MovieResponse } from '../models/movie'
import { MovieDetails } from '../models/movieDetails'
import axios from '../utils/axios'

export const getMovies = (title: string) => {
    return axios.get<MovieResponse>(`/api/movies`, {
        params: { title },
    })
}

export const getMovieDetail = (slug: string) => {
    return axios.get<MovieDetails>(`/api/movie-detail/${slug}`)
}

export const updateMovieFavorite = (slug: string, favorite: boolean) => {
    return axios.post<MovieDetails>(`/api/movie-detail/${slug}`, { favorite })
}
