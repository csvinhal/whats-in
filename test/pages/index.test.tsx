import { fireEvent, waitFor } from '@testing-library/dom'
import { AxiosResponse } from 'axios'
import * as moviesApi from '../../config/moviesApi'
import { MovieResponse } from '../../models/movie'
import { Home } from '../../pages/index'
import { render } from '../testUtils'
import moviesJson from './__mocks__/movies.json'

const axiosResponse: AxiosResponse<MovieResponse> = {
    data: moviesJson,
    status: 200,
    statusText: 'OK',
    config: {},
    headers: {},
}

describe('Home page', () => {
    it('should matches snapshot', () => {
        const { asFragment } = render(<Home />, {})
        expect(asFragment()).toMatchSnapshot()
    })

    it('should show movies', async () => {
        jest.spyOn(moviesApi, 'getMovies').mockImplementation(() =>
            Promise.resolve(axiosResponse),
        )

        const { container, getByLabelText } = render(<Home />, {})
        const input = getByLabelText('Search movies...')
        expect(input).toBeTruthy()

        fireEvent.change(input, { target: { value: 'What' } })

        await waitFor(() => expect(moviesApi.getMovies).toHaveBeenCalled())

        expect(container.querySelectorAll('a.movie-card')).toHaveLength(10)
    })
})
