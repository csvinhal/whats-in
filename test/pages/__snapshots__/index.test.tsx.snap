// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page should matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="jsx-2234572585 layout"
  >
    <header
      class="jsx-2234572585 container mb-3"
    >
      <nav
        class="jsx-2234572585"
      >
        <a
          aria-label="What's in"
          class="jsx-2234572585"
          href="/"
        >
          <div
            style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
          >
            <div
              style="box-sizing: border-box; display: block; max-width: 100%;"
            >
              <img
                alt=""
                aria-hidden="true"
                role="presentation"
                src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTExIiBoZWlnaHQ9IjMyIiB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZlcnNpb249IjEuMSIvPg=="
                style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
              />
            </div>
            <img
              alt="Logo"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="visibility: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
          </div>
        </a>
      </nav>
    </header>
    <main
      class="jsx-2234572585"
    >
      <div
        class="jsx-1504584403 grid"
      >
        <div
          class="jsx-2249794958 grid-item grid-item-xs-true"
        >
          <div
            class="jsx-2629231181 container__search mb-4  container__search--is-empty-state"
          >
            <form
              class="jsx-4126210830 search-bar"
            >
              <svg
                class="search-bar__icon"
                height="1rem"
                version="1.1"
                viewBox="0 0 16 16"
                width="1rem"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g
                  fill="none"
                  fill-rule="evenodd"
                  id="Styleguide-presentation"
                  stroke="none"
                  stroke-width="1"
                >
                  <g
                    id="8.icons"
                    transform="translate(-674.000000, -348.000000)"
                  >
                    <g
                      id="icon__magnifier"
                      transform="translate(674.000000, 348.000000)"
                    >
                      <rect
                        height="16"
                        id="icon__search"
                        width="16"
                        x="0"
                        y="0"
                      />
                      <path
                        d="M14.6666667,14.6666667 L10.6666667,10.6666667"
                        id="Line"
                        stroke="currentColor"
                        stroke-linecap="square"
                        stroke-width="2"
                      />
                      <circle
                        cx="6.66666667"
                        cy="6.66666667"
                        id="Oval-4"
                        r="5.33333333"
                        stroke="currentColor"
                        stroke-width="2"
                      />
                    </g>
                  </g>
                </g>
              </svg>
              <label
                class="jsx-4126210830 search-bar__label"
                for="search-bar"
              >
                Search movies...
              </label>
              <input
                class="jsx-4126210830 search-bar__input"
                id="search-bar"
                placeholder="Search movies..."
              />
            </form>
          </div>
        </div>
      </div>
      <div
        class="jsx-1504584403 grid"
      >
        <div
          class="jsx-2249794958 grid-item grid-item-xs-true"
        >
          <div
            class="jsx-4159469862 empty-state"
          >
            <div
              class="jsx-4159469862 empty-state__img"
            >
              <div
                style="display: block; overflow: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; margin: 0px;"
              >
                <img
                  alt="Empty state image"
                  decoding="async"
                  src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                  style="visibility: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: contain;"
                />
              </div>
            </div>
            <p
              class="jsx-4266605292 typography empty-state__title"
            >
              Don't know what to search?
            </p>
            <p
              class="jsx-1346695943 typography"
            >
              Here's an offer you can't refuse
            </p>
          </div>
        </div>
      </div>
    </main>
  </div>
</DocumentFragment>
`;
